server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java

```diff
@@ -0,0 +1,170 @@
+     *
+     * @param currentState current cluster state
+     * @param actions ordered list of modifications to perform
+     * @return resulting cluster state after all modifications have been performed
+     */
+    static ClusterState modifyDataStream(
+        ClusterState currentState,
+        Iterable<DataStreamAction> actions,
+        Function<IndexMetadata, MapperService> mapperSupplier
+    ) {
+        Metadata updatedMetadata = currentState.metadata();
+
+        for (var action : actions) {
+            Metadata.Builder builder = Metadata.builder(updatedMetadata);
+            if (action instanceof DataStreamAction.AddBackingIndex) {
+                addBackingIndex(
+                    updatedMetadata,
+                    builder,
+                    mapperSupplier,
+                    action.getDataStream(),
+                    ((DataStreamAction.AddBackingIndex) action).getIndex()
+                );
+            } else if (action instanceof DataStreamAction.RemoveBackingIndex) {
+                removeBackingIndex(
+                    updatedMetadata,
+                    builder,
+                    action.getDataStream(),
+                    ((DataStreamAction.RemoveBackingIndex) action).getIndex()
+                );
+            } else {
+                throw new IllegalStateException("unsupported data stream action type [" + action.getClass().getName() + "]");
+            }
+            updatedMetadata = builder.build();
+        }
+
+        return ClusterState.builder(currentState).metadata(updatedMetadata).build();
+    }
+
+    private static void addBackingIndex(
+        Metadata metadata,
+        Metadata.Builder builder,
+        Function<IndexMetadata, MapperService> mapperSupplier,
+        String dataStreamName,
+        String indexName
+    ) {
+        var dataStream = validateDataStream(metadata, dataStreamName);
+        var index = validateIndex(metadata, indexName);
+
+        try {
+            MetadataMigrateToDataStreamService.prepareBackingIndex(
+                builder,
+                index.getWriteIndex(),
+                dataStreamName,
+                mapperSupplier,
+                false);
+        } catch (IOException e) {
+            throw new IllegalArgumentException("unable to prepare backing index", e);
+        }
+
+        // add index to data stream
+        builder.put(dataStream.getDataStream().addBackingIndex(metadata, index.getWriteIndex().getIndex()));
+    }
+
+    private static void removeBackingIndex(Metadata metadata, Metadata.Builder builder, String dataStreamName, String indexName) {
+        var dataStream = validateDataStream(metadata, dataStreamName);
+        var index = validateIndex(metadata, indexName);
+
+        builder.put(dataStream.getDataStream().removeBackingIndex(index.getWriteIndex().getIndex()));
+
+        // remove or disable DataStreamTimestampFieldMapper
+        // TODO: is this necessary or desirable?
```

Comment:
@martijnvg, do you think we should either remove or disable the `DataStreamTimestampFieldMapper` on indices that are removed from data streams?

Reply:
I think it is ok to keep this meta field as is?  If an index is removed from a data stream then this means that they always need to be providing a valid `@timestamp` field and value for new documents. Which kind of makes sense given the type of data that data streams hold.